pkg/util/admission/granter.go

```diff
@@ -1878,9 +1941,186 @@ func (*ioLoadListener) adjustTokensInner(
+	//
+	// Since there is some interaction noted in bullet 1, and because it
+	// simplifies the admission control token behavior, we use flush tokens in
+	// an identical manner as compaction tokens -- to be consumed by all data
+	// flowing into L0. Some of this conservative choice will be compensated for
+	// by flushUtilTargetFraction (when the mix of ingestion and actual flushes
+	// are stable). Another thing to note is that compactions out of L0 are
+	// typically the more persistent bottleneck than flushes for the following
+	// reason:
+	// There is a dedicated flush thread. In comparison, with a write amp of W
+	// and maximum compaction concurrency of C, only approximately C/W threads
+	// are working on reading from L0 to compact out of L0. Since W can be 20+
+	// and C defaults to 3 (we plan to dynamically adjust C but one can expect C
+	// to be <= 10), C/W < 1. So the main reason we are considering flush tokens
+	// is transient flush bottlenecks, and workloads where W is small.
+
+	// Compute flush utilization for this interval. A very low flush utilization
+	// will cause flush tokens to be unlimited.
+	intFlushUtilization := float64(0)
+	if im.Flush.WriteThroughput.WorkDuration > 0 {
+		intFlushUtilization = float64(im.Flush.WriteThroughput.WorkDuration) /
+			float64(im.Flush.WriteThroughput.WorkDuration+im.Flush.WriteThroughput.IdleDuration)
+	}
+	// Compute flush tokens for this interval that would cause 100% utilization.
+	intFlushTokens := float64(im.Flush.WriteThroughput.PeakRate()) * adjustmentInterval
+	intWriteStalls := cumWriteStallCount - prev.cumWriteStallCount
+
+	// Ensure flushUtilTargetFraction is in the configured bounds. This also
+	// does lazy initialization.
+	const maxFlushUtilTargetFraction = 1.5
+	flushUtilTargetFraction := prev.flushUtilTargetFraction
+	minFlushUtilTargetFraction := float64(minFlushUtilizationPercentage) / 100
+	if flushUtilTargetFraction == 0 {
+		// Initialization: use the maximum configured fraction.
+		flushUtilTargetFraction = minFlushUtilTargetFraction
+		if flushUtilTargetFraction < maxFlushUtilTargetFraction {
+			flushUtilTargetFraction = maxFlushUtilTargetFraction
+		}
+	} else if flushUtilTargetFraction < minFlushUtilTargetFraction {
+		// The min can be changed in a running system, so we bump up to conform to
+		// the min.
+		flushUtilTargetFraction = minFlushUtilTargetFraction
+	}
+	numFlushTokens := int64(unlimitedTokens)
+	// doLogFlush becomes true if something interesting is done here.
+	doLogFlush := false
+	smoothedNumFlushTokens := prev.smoothedNumFlushTokens
+	const flushUtilIgnoreThreshold = 0.05
+	if intFlushUtilization > flushUtilIgnoreThreshold {
+		if smoothedNumFlushTokens == 0 {
+			// Initialization.
+			smoothedNumFlushTokens = intFlushTokens
+		} else {
+			smoothedNumFlushTokens = alpha*intFlushTokens + (1-alpha)*prev.smoothedNumFlushTokens
+		}
+		const flushUtilTargetFractionIncrement = 0.025
+		// Have we used >= 90% of the tokens.
```

Comment:
// Have we used, over the last (15s) cycle, more than 90% of the tokens we would give out for the next cycle?